### PR TITLE
Ts rewrite 2

### PR DESCRIPTION
**Pull request by itajaja**
@ethanve and @taion, please have a look at this. This is much more breaking than Ethan's proposal, so I'd like to have some strong feedback please. Turns out, the implementation really leverages some quirks of flow, which makes it hard/awkward or downright impossible to type this thing right. the main difference is that now `NodeType` doesn't come from a factory, but it's an actual static type. We could make the change less breaking by still having `createNodeType` return the NodeType and expose it only through that, but that feels sort of dumb... I'll add comments on other parts

### CODE REVIEW

**Comment by ethanve** [comment]
Can you PR off my branch to see the diff?

**Comment by itajaja** [comment]
sorry - forgot to push last commit, wait a sec

**Comment by itajaja** [comment]
done, have a look

File: test/NodeType.test.ts [hunk 1]
```diff
@@ -7,16 +7,31 @@ import {
...
+  // let context: any;
+  // let NodeType;
+  // let createResolve;
+  // const { NodeType, createResolve } = rest;
+
+  const context = {
+    httpApi: new TestHttpApi(),
+  };
+
+  setup({ localIdFieldMode: 'deprecated' });
```
Comment (jquense) [inline test/NodeType.test.ts]: this seems like a worse API, what's the reason the old one doesn't work in TS?

File: test/NodeType.test.ts [hunk 2]
```diff
@@ -45,41 +60,33 @@ describe('NodeType', () => {
...
+    // ({ NodeType, createResolve } = rest);
 
     class WidgetResource extends TestHttpResource {
       getFoo() {
         return this.api.foo();
       }
     }
 
-    const User = new NodeType({
+    const User: NodeType<WidgetResource> = new NodeType<WidgetResource>({
```
Comment (jquense) [inline test/NodeType.test.ts]: this _should_ be inferable I think? if createResource returns T?

File: test/NodeType.test.ts (hunk 1 above)
Comment (itajaja) [reply test/NodeType.test.ts]: the api _could actually_ be the same, but, still we'd need a static class definition. and once you do that, the old APIs are sort of a lie. The reason why we need a class def is because there are a couple of overlapping problems:

1) [TS does not infer the type of a value if the value is used in the constructor](https://www.typescriptlang.org/play/#code/MYGwhgzhAEBiD29oG8CwAoa1jwHYQBcAnAV2APiIAoAzE3YALmioEpoBeAPmjFwE92yAL4ZR6DDnwFoAS07RcAUwDucRFTaces1gG4gA). This happens when you use `X.getResource`, so you have to do `const Widget: NodeType = new NodeType`. which brings me to the second problem...
2) if a class is generated by a factory, *even if* it's a class, it's not treated as a type in TS, which means that, in order to do the above, you'd have to do weird contorsions like:
  ```ts
  const Widget: NodeTypeInterface = new NodeType()

  // the code below is not valid:
  const Widget: NodeType = new NodeType()
  // because NodeType is not a type. you could do `typeof NodeType`, but then you give up the generics
  ```

Comment (itajaja) [reply test/NodeType.test.ts]: on the bright side, having a singleton makes some things easier and avoids having to pass extra stuff around/using closures

File: test/NodeType.test.ts (hunk 2 above)
Comment (itajaja) [reply test/NodeType.test.ts]: see comment above

Comment (jquense) [reply test/NodeType.test.ts]: not sure i understand isn't this: https://www.typescriptlang.org/play/?ssl=7&ssc=28&pln=1&pc=1#code/MYGwhgzhAEBiD28A8AVAfNA3gWAFDWmHgDsIAXAJwFdgz4KAKAMyuOAC5oGBKaAXgwpeOXHgIBfPJNG4ipMtACW-aMQCmAdziIGPfhgAs3ANxA

effectively the same thing but `i` here is properly inferred as `Foo<number>`

Comment (itajaja) [reply test/NodeType.test.ts]: I don't think your example applies. the constructor is not referencing itself. what am I missing?

Comment (itajaja) [reply test/NodeType.test.ts]: too make it more relevant: 

```ts
const Study = new NodeType({
  fields: {
    resolve: () => Study.getResource().get()
  }
});
```

Comment (itajaja) [reply test/NodeType.test.ts]: err... https://www.typescriptlang.org/play/#code/MYGwhgzhAEByD2ATApgFQJ4AdnQN4FgAoaaYeAOwgBcAnAV2CvhoAoAzAS2REQgC5oAJWRkaiADzUaHcgHMANHmg1kEeCABuyASwCU0ALwA+aGHLpoAXyP6ChIiUtEH0WcirC1dGsG2nzRE72hGSUVNAAylR0iBYG0OTIAO5wSGhYyCx2JJzcvALZJMqq6lo6+saR0bEAdG4eJd6+enXuei5BlroA3EA

Comment (jquense) [reply test/NodeType.test.ts]: ugh, that's so dumb...

Comment (jquense) [reply test/NodeType.test.ts]: seems like we should change the API _here_ then, if we are going to have re annotate every Node then there isn't much value to types here.

Comment (itajaja) [reply test/NodeType.test.ts]: I agree with that! but also, I think the factory is not that great...

**Comment by ethanve** [comment]
I know we discussed this offline but adding here b/c I don't remember the answer 😆. 

Maybe we shouldn't expose the singleton at all and have any configuration needed be passed directly to `NodeType`? The only option as of now is `localIdFieldMode` and the consumer of that is `NodeType` itself (config blindly passes it down without doing anything with it). We could just add that to `NodeTypeConfig` when instantiating a node and instead `setup` is just inlined in the config. 

Not exactly scaleable if we add more options but we can maybe solve that by subclassing?

```ts
// src/widget.ts
const Widget: ... = new NodeType({..., localIdFieldMode: 'omit', ...})

// or subclass if necessary 

// src/NodeType.ts
class NodeType extends NodeTypeBase {
  constructor(config) {
    super({...config, localIdFieldMode: 'omit'})
  }
}
```

**Comment by itajaja** [comment]
@ethanve the problem is that there are other singleton objects that need to be shared: https://github.com/4Catalyzer/graphql-node-resource/pull/312/files#diff-7e82dc7533620e594f349d3ea425b5b4R21 like nodeInterface, the types' and resources' maps, etc

**Comment by jquense** [comment]
I thought most of the value of the factory was to get types out with the right `Context` generic already provided. I'm guessing that doesn't work here? I don't actually mind a singleton defaults() method that much

**Comment by itajaja** [comment]
> I thought most of the value of the factory was to get types out with the right Context generic already provided

I think for these kind of things, since the Context is "global" across your application, the best thing to do is to use TS interface extension to type your Context appropriately. This is what I have in the QSI pr I am working on:

```
type Context = Omit<
  import('@4c/graphql-node-resource').HttpContext,
  'httpApi'
> & {
  httpApi: import('../src/api/HttpApi').default;
};
```

**Comment by ethanve** [comment]
> @ethanve the problem is that there are other singleton objects that need to be shared: https://github.com/4Catalyzer/graphql-node-resource/pull/312/files#diff-7e82dc7533620e594f349d3ea425b5b4R21 like nodeInterface, the types' and resources' maps, etc

You would still have the config but instantiate it immediately

**Comment by itajaja** [comment]
> You would still have the config but instantiate it immediately

you could, I am afraid that doing it so implicitly might make tests harder

File: src/resources/Resource.ts [hunk 3]
```diff
@@ -1,16 +1,11 @@
...
   constructor(context: TContext) {
     this.context = context;
   }
 
-  abstract get<T>(
-    _id: string,
-  ): (T | null | undefined) | Promise<T | null | undefined>;
-
-  // TODO: Once we have a concept of a model, this should go on there.
-  makeId({ id }: any): string {
```
Comment (taion) [inline src/resources/Resource.ts]: why did we delete this?

File: src/resources/HttpResource.ts [hunk 4]
```diff
@@ -13,20 +15,21 @@ export type HttpResourceOptions = {
...
 
-  constructor(context: HttpContext<TApi>, { endpoint }: HttpResourceOptions) {
+  constructor(context: TContext, { endpoint }: HttpResourceOptions) {
     super(context);
 
     this.api = context.httpApi;
     this._endpoint = endpoint;
   }
 
-  get<T>(id?: string): (T | null | undefined) | Promise<T | null | undefined> {
```
Comment (taion) [inline src/resources/HttpResource.ts]: should this no longer be parametrized on `<T = Obj>`?

is that just useless because TS doesn't do this sort of inference? if so, then why do we keep this on `HttpApi`?

File: src/config.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import { GraphQLFieldConfig, GraphQLInterfaceType } from 'graphql';
+// eslint-disable-next-line max-classes-per-file
```
Comment (taion) [inline src/config.ts]: this can't possibly be doing anything in this file that has no classes

File: src/config.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  typesManager: Map<string, NodeType<Resource<Context>, any>>;
+  nodeInterface: GraphQLInterfaceType;
+};
+
+export const config: Readonly<Config> = {} as any;
+
+function createConfig({
+  localIdFieldMode = 'omit',
+}: Pick<Config, 'localIdFieldMode'>): Config {
+  const typesManager = new Map<string, NodeType<Resource<Context>, any>>();
```
Comment (taion) [inline src/config.ts]:
```suggestion
  const nodeTypesByName = new Map<string, NodeType<Resource<Context>, any>>();
```

maybe?

File: src/config.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+  const { nodeInterface, nodeField, nodesField } = nodeDefinitions<Context>(
+    async (globalId, context) => {
+      const { type, id } = fromGlobalId(globalId);
+      const resolvesType = typesManager.get(type);
```
Comment (taion) [inline src/config.ts]:
```suggestion
      const nodeType = nodeTypeByName.get(type);
```

Comment (taion) [reply src/config.ts]: or `resolvedType`

File: src/config.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    localIdFieldMode,
+    nodeInterface,
+    nodeField,
+    nodesField,
+  };
+}
+
+export function needsSetup() {
+  return Object.keys(config).length === 0;
+}
```
Comment (taion) [inline src/config.ts]: this seems slightly awkward. could you just define e.g. `config` as `Config | null`, and have a `getConfig` method that always does the check and returns `Config`?

File: src/types/ResourceManager.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import { Context, RESOURCE_CACHE_KEY } from './Context';
+import Resource from '../resources/Resource';
+
+export default class ResourceManager {
```
Comment (taion) [inline src/types/ResourceManager.ts]:
```suggestion
export default class ResourceCache {
```

or `ResourceCacheManager`?

File: src/utils/typing.ts [hunk 10]
```diff
@@ -0,0 +1,5 @@
+export type ValueOrPromise<T> = T | Promise<T>;
```
Comment (taion) [inline src/utils/typing.ts]: graphql-js uses

```js
export type PromiseOrValue<T> = Promise<T> | T;
```

Comment (taion) [reply src/utils/typing.ts]: might as well match what they have :p

File: src/utils/urlJoin.ts [hunk 11]
```diff
@@ -1,5 +1,7 @@
-export default function urlJoin(...parts: Array<any>) {
-  return parts.reduce((r, next) => {
+export default function urlJoin(
+  ...parts: Array<string | undefined | null>
```
Comment (taion) [inline src/utils/urlJoin.ts]: i think these might actually be allowed to include numbers and such

File: src/config.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+export function setup(options: Parameters<typeof createConfig>[0]) {
```
Comment (taion) [inline src/config.ts]: you should not allow calling this twice

File: src/resources/Resource.ts (hunk 3 above)
Comment (itajaja) [reply src/resources/Resource.ts]: this is now on Node, so that it can use TSource from the node

**Comment by taion** [comment]
why is `makeId` on the node type itself now?

**Comment by taion** [comment]
well, losing typing on context... meh 

it looks like you’ve made other api changes to make node type richer. can you summarize what they were,

**Comment by itajaja** [comment]
> well, losing typing on context... meh

you don't lose type, you extend the Context interface locally. I think this makes sense, especially with a singleton pattern, and allows you to carry a generic around.

main changes

- singleton
- createResource instead of Resource/resourceConfig which is not great from a typing standpoint
- createResolve takes an array of fields as second params
- TSource is the "stub" object, by default is `{id: string}` but that can be changed when objects have composite IDs.

File: src/utils/urlJoin.ts (hunk 11 above)
Comment (itajaja) [reply src/utils/urlJoin.ts]: just adding `number`s for now

File: src/utils/typing.ts (hunk 10 above)
Comment (itajaja) [reply src/utils/typing.ts]: actually I am not using it :)

File: src/config.ts (hunk 8 above)
Comment (itajaja) [reply src/config.ts]: sucks to call getConfig all the time

Comment (itajaja) [reply src/config.ts]: I could you make config a proxy that checks that though 😬

File: src/resources/HttpResource.ts (hunk 4 above)
Comment (itajaja) [reply src/resources/HttpResource.ts]: we can add it, however it's just not a lot of value add. I am pretty confident we could drop it from httpApi, but I view it as lower level so it might be good to give the option to specify something

**Comment by itajaja** [comment]
FYI this branch is now in production at qsi 👍

File: src/config.ts (hunk 8 above)
Comment (taion) [reply src/config.ts]: i mean how many times are you accessing the config? it's not that much. and it's never in user code

File: src/types/NodeType.ts [hunk 13]
```diff
@@ -1,356 +1,159 @@
...
-// to manually assign a cache key for the full batch.
-export const RESOURCE_CACHE_KEY = Symbol('resource cache key');
-
-export interface Context {
-  [RESOURCE_CACHE_KEY]: any;
-}
+export interface NodeTypeConfig<R extends Resource, TSource>
+  extends GraphQLObjectTypeConfig<TSource, R['context']> {
+  localIdFieldName?: string | null | undefined;
+  createResource: (contxt: R['context']) => R;
```
Comment (taion) [inline src/types/NodeType.ts]:
```suggestion
  createResource: (context: R['context']) => R;
```

File: src/config.ts (hunk 8 above)
Comment (itajaja) [reply src/config.ts]: true

**Comment by itajaja** [comment]
> since we don't have and can't really get the object type from the resource, but can only go from resource to object type, i think we're better off having the resource have `makeId`.

a couple of reasons why I disagree.

HttpResource simply does not have idea how to make an Id. providing a default with stub id looks odd. HttpResource can by design be used in NodeTypes with different `TSource`, so it follows that HttpResource cannot make assumptions about the shape of the objects that is passed to makeId.
A generic resource never calls or need makeId. the endpoint configuration is already on node, and it makes assumptions about how the id is made, so it makes sense to colocate the two.

**Comment by taion** [comment]
We actually have a number of places in our codebase where we call `this.makeId`. I don't know that there's an absolutely perfect way to do this, but I'd rather minimize how much we extend functionality.

Or at least, I see it as a net hit to code quality that we can't call `resource.get(resource.makeId(...))` and instead have to make sure that we sync up to the correct node type.

None of this is perfect – the resource abstraction is fairly broken in any event. But looking through our code, I'd rather have `makeId` live on the resource.

Can you make that plus the noted changes, and then let's ship this?

**Comment by itajaja** [comment]
> We actually have a number of places in our codebase where we call this.makeId. I don't know that there's an absolutely perfect way to do this, but I'd rather minimize how much we extend functionality.

in these places, can't you do `TheNodeType.makeId`?

> Or at least, I see it as a net hit to code quality that we can't call resource.get(resource.makeId(...)) and instead have to make sure that we sync up to the correct node type.

this is really an arguable point, since 99% of the time the resource comes from a specific node, so there is really no ambiguity

feel free to try putting it on the resource, I'd be fine with it. I couldn't find a way to get a satisfiable level of type safety.

**Comment by taion** [comment]
Is there no other way out? Can we not bind the node type to the resource at all?

**Comment by itajaja** [comment]
what do you mean "bind"?

**Comment by taion** [comment]
I meant something like "be able to access `this.nodeType` on the resource... not important if typing isn't possible

**Comment by itajaja** [comment]
@taion can you take another look

**Comment by ethanve** [comment]
Merging this in to my branch. Lets go!
